Inelastic predictor

lynnmunday opened the pull request:
closes #27429

lynnmunday committed:
corrected computation of effective strain from 3/2 to 2/3 to compute number of substeps and regolded or modified tests

lynnmunday committed:
Adding inelastic strain increment to predict intial strain increment in radial
return mapping and intializing the model around a modified effective trial stress
that accounts for this strain increment.
Breaks Bison
need to fix regular expression in test file.
ref #27429

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2197567/) on 6d12ef8 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27492/site)

This comment will be updated on new commits.

lynnmunday committed:
Address reviewer comments ref #27492

lynnmunday committed:
Changing to strain increment predictor to use the effective strain rate from the previous step to make this work with substepping.  ref #27429

lynnmunday committed:
Documenting effective strain increment predictor closes #27429

lynnmunday commented:
This pr breaks bison.  I will fix it when we merge this.

lynnmunday commented on modules/solid_mechanics/src/materials/RadialReturnStressUpdate.C:
```diff
@@ -250,21 +261,32 @@ RadialReturnStressUpdateTempl<is_ad>::updateState(
                                                   : 0.0;
+  // Initialize models around a stress that uses the previous inelastic strain increment to
+  // predict the final stress state for this step. (Dunne & Petrinic eq 5.16 )
+  // fixme three shear mod is set in comptueStressInitialize
```
_three_shear_modulus is set in RadialReturnStressUpdate::computeStressInitialize.  We then call RadialReturnStressUpdate::computeStressInitialize from all of the derived computeStressInitialize and we check that we called it in the Assert on ln 284. 
 Why don't we just set it in this function before we call computeStressInitialize?  Then make RadialReturnStressUpdate::computeStressInitialize a no-op function and not call it from any derived classes.

dschwen commented on modules/solid_mechanics/doc/content/source/materials/RadialReturnStressUpdate.md:
```diff
@@ -112,6 +112,12 @@ allowed by substepping. A value of $1.0\cdot10^4$ will work for many cases. An e
+
+The scalar inelastic strain increment $\Delta p$ is normally intialized to zero and the constitutive model at the start of each timestep is intialized around a trial stress computed from a fully elastic strain increment predictor.  [!param](/Materials/PowerLawCreepStressUpdate/scale_strain_predictor)  allows a scaled value of the previous timesteps final $\Delta p$ to initialize the current times ...
+
+The scaling factor [!param](/Materials/PowerLawCreepStressUpdate/scale_strain_predictor) ranges between zero and one.  Zero is the default which will not use the previous timestep's $\Delta p$.  The entire previous step's $\Delta p$ is used by setting `scale_strain_predictor`$=1$ which is usually too large of a guess for $\Delta p$ and will result in a constitutive model state where the stress is ...
```
> The entire previous step's $\Delta p$ is used by setting `scale_strain_predictor`$=1$ which is usually too large of a guess for $\Delta p$ and will result in a constitutive model state where the stress is inside the yield surface.  A check is done to ensure the constitutive model state results in a stress inside the yield surface, and if it violates this condition, `scale_strain_predictor` is set to zero,

This is confusing:
1. A `scale_strain_predictor` of 1 is usually **too large** and will result in a stress **inside the yield surface**. (inside = bad?
2. You check to **ensure** the stress is **inside the yield surface. (inside = good?)
3. If the stress violates the condition (inside = bad again?) set scale_strain_predictor to zero (why not progressively cut it in half?)

dschwen commented on modules/solid_mechanics/include/materials/ADViscoplasticityStressUpdate.h:
```diff
@@ -47,7 +47,8 @@ class ADViscoplasticityStressUpdate : public ADViscoplasticityStressUpdateBase,
    */
-  virtual ADReal initialGuess(const ADReal & effective_trial_stress) override;
+  virtual ADReal initialGuess(const ADReal & effective_trial_stress,
+                              const ADReal & effective_strain_inc) const override;
```
```suggestion
                              const ADReal & effective_strain_increment) const override;
```

dschwen replied:
everywhere

dschwen commented on modules/solid_mechanics/src/materials/RadialReturnStressUpdate.C:
```diff
@@ -80,6 +84,12 @@ RadialReturnStressUpdateTempl<is_ad>::RadialReturnStressUpdateTempl(
+    _eff_inelastic_strain_rate_old(this->template getMaterialPropertyOld<Real>(
+        this->_base_name + this->template getParam<std::string>("effective_inelastic_strain_name") +
+        "_rate")),
     _max_inelastic_increment(this->template getParam<Real>("max_inelastic_increment")),
```
just like it's spelled out here

dschwen commented on modules/solid_mechanics/src/materials/RadialReturnStressUpdate.C:
```diff
@@ -250,21 +261,32 @@ RadialReturnStressUpdateTempl<is_ad>::updateState(
 
   // Use Newton iteration to determine the scalar effective inelastic strain increment
-  _effective_inelastic_strain_increment = 0.0;
+  _effective_inelastic_strain_increment = predicted_eff_strain_inc;
```
```suggestion
  _effective_inelastic_strain_increment = predicted_eff_strain_increment;
```

everywhere

lynnmunday replied:
I thought we could use abbreviations for local variables.

lynnmunday committed:
Addressing reviewer comments closes #27429

lynnmunday commented:
@dschwen will you review again and merge and then I'll fix Bison?

lynnmunday commented:
mark as not stale

lynnmunday committed:
adding material properties for material nonlinear convergence and residual.

lynnmunday committed:
Using csvdiff for testing instead of expect_out for strain predictor testing with substepping.
Loosened the material radial return absolute tolerance for the predictor tests to help with csv diffing.

lynnmunday commented:
Technical Debt tracking:  Total time for commit was 60 hours.   Time for new code and testing and documentation was 30 hours.  Time for refactoring and being constrained by other code was 30 hours.  The singleVariableReturnMappingSolution::initialGuess is not being used consistently.  Substepping was also difficult.

lynnmunday commented:
@dschwen will you review this again

dschwen commented on modules/solid_mechanics/src/materials/SingleVariableReturnMappingSolution.C:
```diff
@@ -449,5 +461,14 @@ SingleVariableReturnMappingSolutionTempl<is_ad>::outputIterationSummary(
+    const GenericReal<is_ad> & /*effective_trial_stress*/,
+    const GenericReal<is_ad> & effective_strain_increment)
+{
+  return effective_strain_increment;
```
Meaning

```
template <class T, bool is_ad>
class SingleVariableReturnMappingSolutionTempl : public T
{
```

where `T` is a material base class that your downstream class inherited from. So

```
class MyMaterial: public Material, public SingleVariableReturnMappingSolution
```

would turn into

```
class MyMaterial: public SingleVariableReturnMappingSolution<Material>
```

Thereby `SingleVariableReturnMappingSolution` inserts itself inbetween your class and `Material` and would have access to material properties. That means we can reduce the number of function arguments and move to using more member variables. 

My suggestion would be to start a new return mapping class template for this and progressively migrate to the new design.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2638660/), step Docs: sync website on 2dcfa59 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27492/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2638707/), step Generate coverage on 2dcfa59 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>2df091</th>
      <th colspan=4>#27492 2dcfa5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.00%</td>
      <td>85.02%</td>
      <td>+0.01%</td>
      <td>97.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>28504</td>
      <td>28534</td>
      <td>+30</td>
      <td>39</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>5029</td>
      <td>5029</td>
      <td>-</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27492/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27492/coverage/solid_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27492/coverage/xfem)

</details>



This comment will be updated on new commits.
